# 第03课：列表、文件与循环

本笔记本是课堂讲义。每个知识点包含：理论知识、案例代码、讲解、易错点与练习。综合练习 P1 使用同目录的 [mobile_game_info.csv](mobile_game_info.csv)。课后独立练习见 [chapter03_列表文件与循环_课后练习.ipynb](chapter03_列表文件与循环_课后练习.ipynb)，数据为 [googleplaystore.csv](googleplaystore.csv)。

## 学习目标

1. 理解列表的有序性、可修改性，使用索引、元素赋值、len和切片。
2. 区分整张表、一条记录和单个字段，用嵌套列表定位数据。
3. 理解文件对象、CSV reader和列表的区别，读取并检查课程文件。
4. 解释for循环的语法、变量变化和缩进规则。
5. 使用类型转换与累加器计算平均值，并检查统计口径与结果。

## 学习知识点

| 数据结构 | 文件读取 | 循环与计算 |
| --- | --- | --- |
| 列表的定义与创建 | 文件路径与open | for语法与执行过程 |
| 正负索引、元素修改 | with与文件关闭 | 循环变量与缩进 |
| len、嵌套列表、切片 | csv.reader、list、header/body | float、累加器、平均值 |

## 基础回顾与案例提问

课程数据中“明日方舟”的score字段为文本`"8.2"`。先用已有知识判断：

1. **R.1** `score_text = "8.2"`之后，score_text的类型是什么？`score_text[0]`又是什么类型？
2. **R.2** `score = float(score_text)`会把score_text也改成数值吗？用两个type检查自己的判断。
3. **R.3** `"8.2" + "8.8"`和`8.2 + 8.8`分别进行什么运算？先写预测，再运行核对。

**作答：** 预测：____；依据：____；验证后说明：____。

所有游戏案例均取自[mobile_game_info.csv](mobile_game_info.csv)。开始时从原文件摘取少量字段解释语法，随后直接读取同一文件。使用Python 3；只需标准库csv。让CSV与Notebook处于同一文件夹，从该文件夹启动内核，按顺序运行案例。本课 CSV 可能带有 UTF-8 BOM，打开文件时使用 encoding="utf-8-sig"，表头第一列才会是 id。

In [ ]:
# R.1–R.3: Write and verify your predictions here.

## 1. 列表的定义与创建

### 理论知识

**列表（list）是一种有序、可修改的序列。** 它把多个元素保存在一个对象中，用于表示一组相关数据。

- **有序**：元素保留放入时的位置，可以按位置访问；有序不等于自动按大小排序。
- **可修改**：创建后可以替换其中的元素，不必重新创建整条记录。
- **元素类型不必相同**：一个列表可以同时含字符串和数值；每个元素仍有自己的类型。
- **允许重复**：相同的值可以出现在多个位置，列表不会自动去重。

基本语法：

```python
items = [element1, element2, element3]
empty_items = []
```

这里element1等表示准备放入的值或已有变量，不是必须使用的名称。方括号创建列表，逗号分隔元素；引号决定某个元素是不是字符串。

### 案例：一个列表对象与它的元素

In [1]:
items = ["明日方舟", 8.2, "策略"]
empty_items = []
print(items)
print(type(items))
print(empty_items)

['明日方舟', 8.2, '策略']
<class 'list'>
[]


### 讲解

items整体的类型是list，其中`"明日方舟"`和`"策略"`是字符串，8.2是浮点数。empty_items也是列表，只是还没有元素。

原CSV中的所有字段都是文本。摘取同一条记录的title、score、game_type时，可以先保留文本类型：

In [2]:
one_game = ["明日方舟", "8.2", "策略"]
print(one_game)
print(type(one_game))

['明日方舟', '8.2', '策略']
<class 'list'>


### 易错点与练习

`["明日方舟", "8.2", "策略"]`是三个元素的列表；`"明日方舟,8.2,策略"`是一个字符串。字符串中出现逗号，不会自动创建列表。

1. **K1.1** 根据原记录“人格解体、评分文本8.8、类型角色扮演”，创建三字段列表，打印它及其类型。
2. **K1.2** 判断`["策略", "策略"]`是否允许；它是否会自动变成只有一项？创建并观察，说明列表的哪项性质解释了结果。

**作答：** 类型依据：____；重复元素判断及验证：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

## 2. 列表索引：按位置访问元素

### 理论知识与语法

**索引（index）是元素的位置编号。** 使用`items[index]`取出一个元素，原列表不会因为取值而改变。

- 非负索引从左往右：0表示第一项，1表示第二项。
- 负索引从右往左：-1表示最后一项，-2表示倒数第二项。
- 长度为n的非空列表，合法索引范围是`0`到`n-1`，或`-n`到`-1`。
- 索引必须是整数形式；`items[1]`与`items["1"]`不是同一种访问。
- 访问结果的类型由被选元素决定，索引操作不会自动转换类型。

以原记录摘取的三字段列表为例：

| 元素 | 明日方舟 | 8.2（文本） | 策略 |
| --- | --- | --- | --- |
| 非负索引 | 0 | 1 | 2 |
| 负索引 | -3 | -2 | -1 |

### 案例：位置与取值

In [3]:
one_game = ["明日方舟", "8.2", "策略"]
print(one_game[0])
print(one_game[-1])
print(one_game[1], type(one_game[1]))

明日方舟
策略
8.2 <class 'str'>


### 讲解

第一个print取名称，第二个取最后的类型字段，第三个取评分文本并检查类型。评分看起来是数字，但类型仍为str。

换到原文件七字段记录时，字段位置由当前表头决定：title是索引1，score是索引2，game_type是索引3。下面仅摘出原文件第一条的前三个字段，说明“索引1”不会永久代表评分。

In [4]:
record_start = ["206369", "人格解体", "8.8"]
print("Title:", record_start[1])
print("Score:", record_start[2])

Title: 人格解体
Score: 8.8


### 易错点与练习

超出范围会产生IndexError；合法索引也可能选错字段。排查时先写出字段顺序，再检查位置。

1. **K2.1** 对`["一念逍遥", "6.5", "角色扮演"]`，用正索引和负索引各取一次评分，说明两种写法为什么应选中同一元素。
2. **K2.2** 对该三元素列表，用索引3取最后元素会发生什么？预测错误，写出修正表达式；不要把故障语句保留在正常运行流程中。

**作答：** 两种索引的依据：____；错误与修正说明：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 3. 修改列表元素：元素赋值与变量重新赋值

### 理论知识与语法

列表的**可修改性**指可以改变已有位置的元素。基本语法是：

```python
items[index] = new_value
```

先确定items所指的列表，再找到index位置，用new_value替换该位置的旧元素。其他位置保持不变，单元素替换不改变列表长度；被替换元素的类型也可以改变。

这与`items = new_list`不同：后者让变量名指向另一个列表。`items[index]`作为赋值目标时表示“修改这个位置”；仅出现在print中时表示“取出这个位置的值”。

### 案例：把评分文本转换后放回原位置

In [5]:
editable_record = ["明日方舟", "8.2", "策略"]
print(editable_record[1], type(editable_record[1]))
editable_record[1] = float(editable_record[1])
print(editable_record[1], type(editable_record[1]))
print(editable_record)

8.2 <class 'str'>
8.2 <class 'float'>
['明日方舟', 8.2, '策略']


### 代码讲解与数据案例

右侧`float(editable_record[1])`先把文本转成浮点数8.2，再赋回同一位置。因此评分的数值含义没变，但类型从str变为float；名称和类型字段保持不变。

原始CSV的内容没有被改写。Python中的内存对象与硬盘文件是两个层次：修改列表不是保存文件。下面重新用原记录建立变量，可以再次得到评分文本。

In [6]:
original_record = ["明日方舟", "8.2", "策略"]
print(original_record[1], type(original_record[1]))

8.2 <class 'str'>


### 易错点与练习

`items[3] = value`不能给三个元素的列表添加第四项，会越界。单个元素赋值只能替换已存在的位置，本节不使用它扩展列表。

1. **K3.1** 用原记录`["人格解体", "8.8", "角色扮演"]`建立change_record，将评分转为float并放回原位置；打印整个列表及评分类型，说明改变了什么。
2. **K3.2** 比较`change_record[1] = 8.8`和`change_record = 8.8`：哪一种会让变量不再指向列表？写出各自含义，必要时用独立变量验证。

**作答：** 修改前后差异：____；两种赋值的区别：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

## 4. len：长度不是最后一个索引

### 理论知识与语法

`len(object)`是内置函数，返回对象包含的元素数量，结果为整数。对列表，它数最外层元素；对字符串，它数该字符串中的字符。本节主要用它区分记录的字段数和表格的记录数。

```python
number_of_items = len(items)
```

非空列表有n项，最后一个非负索引是n-1。空列表长度为0，没有可以访问的第一个或最后一个元素。因此“长度是多少”与“最后位置在哪里”是两个问题。

### 代码与数据案例

In [7]:
one_game = ["明日方舟", "8.2", "策略"]
print("Field count:", len(one_game))
print("Last nonnegative index:", len(one_game) - 1)
print("Last element:", one_game[len(one_game) - 1])
print("Empty list length:", len([]))

Field count: 3
Last nonnegative index: 2
Last element: 策略
Empty list length: 0


### 代码讲解与易错点

三个字段的数量是3，最后非负索引为2，所取元素为策略。len([])为0，但不能因此访问空列表的索引0。

不要把`len(one_game)`当成“这张表有几个游戏”：one_game是一条记录，len数的是它的字段。函数返回的数字含义取决于传入的对象。

### 练习

1. **K4.1** 原文件前三个名称是人格解体、一念逍遥、明日方舟。将它们放入titles列表，用len得到数量，再根据长度取最后一个名称。
2. **K4.2** `len(["明日方舟", "8.2", "策略"])`与`len("明日方舟")`各在数什么？先预测结果，再运行，解释两个参数的类型差异。

**作答：** 数量和索引关系：____；两个len的含义与验证：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

## 5. 嵌套列表与二维索引

### 理论知识与语法

当列表的某个元素本身也是列表时，就形成嵌套结构。若外层每个元素都代表一条记录，便可以用“列表的列表”表示表格。

```python
table = [row1, row2, row3]
value = table[row_index][column_index]
```

`table[row_index]`先返回一个内层列表，再对该列表用`[column_index]`选字段。执行过程从左往右，不是同时在一个平面上寻找两个数字。

- table是整张表；table[i]是一条记录；table[i][j]是单个字段。
- len(table)数行；len(table[i])数第i行的字段。
- Python允许内层列表长度不同，不会自动检查它是否为规则表格。使用固定列之前，应确认输入有一致的字段结构。

### 代码：把一次连续访问拆成两步

In [8]:
small_table = [
    ["人格解体", "8.8", "角色扮演"],
    ["一念逍遥", "6.5", "角色扮演"]
]
selected_row = small_table[1]
selected_score = selected_row[1]
print(selected_row)
print(selected_score)
print(small_table[1][1])

['一念逍遥', '6.5', '角色扮演']
6.5
6.5


### 讲解与原始数据案例

第一步得到第二条记录`['一念逍遥', '6.5', '角色扮演']`，第二步得到评分文本'6.5'；连续索引得到相同值。

下面从原文件前五条记录摘出title、score、game_type，保持原顺序和原文本。用它建立更完整的小表：

In [9]:
sample = [
    ["人格解体", "8.8", "角色扮演"],
    ["一念逍遥", "6.5", "角色扮演"],
    ["明日方舟", "8.2", "策略"],
    ["2048清", "9.9", "益智"],
    ["炉石传说", "7", "策略"]
]
print("Third title:", sample[2][0])
print("Third score:", sample[2][1])
print("Rows and fields:", len(sample), len(sample[0]))

Third title: 明日方舟
Third score: 8.2
Rows and fields: 5 3


### 结果、易错点与练习

第三条名称为明日方舟，评分文本为'8.2'；表有5行，每行3字段。`sample[1][2]`与`sample[2][1]`选中不同的位置，不能互换。

1. **K5.1** 从sample取出2048清的评分，分别写两步访问和连续索引，并解释两个索引各选什么。
2. **K5.2** 判断`sample[4]`和`sample[4][0]`的类型及含义，先写预测再打印验证。

**作答：** 访问步骤：____；对象层级与类型：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

## 6. 列表切片：取出连续的一段

### 理论知识与语法

索引取一个元素，**切片（slice）取一段元素组成的列表**。本节使用连续切片：

```python
part = items[start:end]
```

规则是**包含start，不包含end**。省略start表示从开头开始，省略end表示到结尾。

| 写法 | 作用 |
| --- | --- |
| items[1:3] | 取索引1、2处的元素 |
| items[:2] | 取前两个元素 |
| items[1:] | 跳过第一个，保留其余元素 |

对列表切片会得到一个新的外层列表，原列表长度和顺序不变。本章只用它选取记录，不通过切片修改内层记录；不展开复制深度问题。

### 比较索引和切片的返回类型

In [10]:
titles = ["人格解体", "一念逍遥", "明日方舟", "2048清", "炉石传说"]
print(titles[0], type(titles[0]))
print(titles[:1], type(titles[:1]))
print(titles[1:3])

人格解体 <class 'str'>
['人格解体'] <class 'list'>
['一念逍遥', '明日方舟']


### 讲解与数据案例

titles[0]得到一个名称字符串，titles[:1]得到只装了一个名称的列表；都“涉及第一个名称”，但返回对象不同。titles[1:3]取索引1、2，所以得到一念逍遥和明日方舟，不包括索引3。

sample已经是五条原始记录的简化表。切片既能截取名称列表，也能截取记录列表：

In [11]:
first_two_records = sample[:2]
remaining_records = sample[1:]
print(first_two_records)
print("Remaining record count:", len(remaining_records))
print("Original record count:", len(sample))

[['人格解体', '8.8', '角色扮演'], ['一念逍遥', '6.5', '角色扮演']]
Remaining record count: 4
Original record count: 5


### 易错点与练习

`sample[1:]`不会删除sample的第一行，而是返回另一个列表；要使用结果，需要接收它或直接处理表达式。切片越过右端会取到结尾，不像访问不存在的单个索引那样越界；例如短列表取[:5]不会自动补足五项。

1. **K6.1** 从sample取出第二、第三、第四条记录，写出切片并说明终点为何这样写。
2. **K6.2** 先预测sample[:2]与sample[2]各包含什么层级的对象，然后分别查看；用一句话说明“取前两条”和“取第三条”的区别。

**作答：** 切片边界解释：____；两个结果的区别：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

## 7. open与with：文件对象及其使用范围

### 理论知识与语法

文件保存在磁盘上，列表存在于程序内存中。**open返回文件对象**，让程序能读取文件内容；文件对象不是已经读好的表格。

读取本章CSV采用：

```python
with open(path, "r", encoding="utf-8-sig", newline="") as file:
    # 在这里使用打开的文件
    print(type(file))
```

| 部分 | 含义 |
| --- | --- |
| path | 字符串形式的文件路径 |
| "r" | 以读取模式打开，不写入文件 |
| encoding="utf-8-sig" | 指定解码文字使用UTF-8 |
| newline="" | 让CSV读取工具按其规则处理换行 |
| as file | 将文件对象绑定给变量file |
| with后的缩进 | 文件保持打开、供本段代码使用的范围 |

with会在离开代码块时关闭文件，包括代码因异常退出该块的情况。file只是变量名，不是关键字。这里理解读取模式即可，不枚举其他文件模式。

若文件以 UTF-8 BOM 开头，用 encoding="utf-8" 读取时，表头第一列可能变成带不可见前缀的 id。本章统一使用 encoding="utf-8-sig"，读取时去掉 BOM。

**相对路径**以程序当前工作目录为起点，不保证总是以点击打开的文件位置为起点。本章约定从Notebook与CSV所在文件夹启动内核，这时文件名本身就是可用的相对路径。

### 代码

In [12]:
with open("mobile_game_info.csv", "r", encoding="utf-8-sig", newline="") as file:
    print(type(file))
print("The with block has ended.")

<class '_io.TextIOWrapper'>
The with block has ended.


### 代码讲解与易错点

缩进的print显示文本文件对象的类型；最后一句没有缩进，执行到它时文件已经关闭。此例只打开并检查对象，**没有把CSV读成列表**。

FileNotFoundError通常要求检查文件名、扩展名和工作目录；中文解码出错时要核对实际编码与读取编码。不要把“打开成功”当成“字段已经读对”。

### 练习

1. **K7.1** 假设内核工作目录就是Notebook所在目录，而mobile_game_info.csv已移动到其data子文件夹。应使用什么相对路径？写出open语句
2. **K7.2** 在上面的with例子中，哪一行执行时文件仍然打开？哪一行执行时已经关闭？说明缩进在这里表达什么--缩进表示代码属于哪个代码块。

**作答：** 路径及理由：____；两行的状态与依据：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

缩进表示代码属于哪个代码块。这里缩进的语句属于 with 内部；恢复到与 with 相同的缩进层级，就表示已经离开这个代码块。

用 `file.closed` 验证如下。


In [13]:
with open("mobile_game_info.csv", "r", encoding="utf-8-sig", newline="") as file:
    print(file.closed)  

print(file.closed)   

False
True


## 8. csv.reader与list：解析记录和收集记录

### 理论知识与语法

**模块（module）提供可以使用的功能。** 本章用`import csv`加载标准库CSV工具，不需要安装额外的数据分析库。

```python
reader = csv.reader(file)
rows = list(reader)
```

这两行解决不同问题：

1. csv.reader读取CSV格式的记录，按分隔符、引号等规则解析字段；每次提供一条记录，记录是字符串组成的列表。
2. list(reader)把reader提供的记录收集起来，形成外层列表，才能方便地用rows[i][j]访问。

reader的读取位置会随着读取向后推进；同一个reader被完整收集后，不会自动回到文件开头。本章需要重读时重新打开文件，不把反复调用list(reader)当作重新读取。

CSV默认读取的字段是str，文件中的8.8不会自动变成float。CSV解析也不等于简单按每个逗号切割文本：被引号包住的字段可以含逗号，所以本章使用csv.reader而不是自己拆字符串。

### 代码：依次观察三种对象

In [14]:
import csv

with open("mobile_game_info.csv", "r", encoding="utf-8-sig", newline="") as file:
    reader = csv.reader(file)
    rows = list(reader)
    print("File object:", type(file))
    print("CSV reader:", type(reader))
    print("Collected table:", type(rows))


File object: <class '_io.TextIOWrapper'>
CSV reader: <class '_csv.reader'>
Collected table: <class 'list'>


### 代码讲解

file用于读取，reader负责按CSV格式提供记录，rows才是列表的列表。收集放在with内部完成；文件关闭后仍可使用已存入内存的rows。

读取完整文件后，先查看第一行和下一行，辨认格式，不急着统计：

In [15]:
print("First file row:", rows[0])
print("Second file row:", rows[1])
print("Score field:", rows[1][2], type(rows[1][2]))

First file row: ['id', 'title', 'score', 'game_type', 'pulisher', 'download', 'isAndroid']
Second file row: ['206369', '人格解体', '8.8', '角色扮演', '厂商:\xa0喵法自然工作室', '71069', 'TRUE']
Score field: 8.8 <class 'str'>


### 结果、易错点与练习

第一行是列名，第二行是人格解体的记录，评分字段为str。print通常不显示单个字符串的引号，所以不能凭输出外观判断类型。

1. **K8.1** 将“文件对象”“CSV reader”“嵌套列表”与open、csv.reader、list三步对应，说明哪一步之后能使用rows[1][2]。
2. **K8.2** 同学先运行`first = list(reader)`读完所有记录，再运行`second = list(reader)`，认为second也是同一张表。这个判断合理吗？解释reader的位置变化，并说明怎样真正重读。

**作答：** 对应关系：____；两次收集的判断及原因：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

## 9. 表头与数据行：header、body的含义

### 理论知识与语法

**表头（header）描述字段，数据行（body中的记录）提供观测值。** CSV并不强制所有文件都有表头；本章文件已经检查，第一行确实是列名，因此可以按下面方式分开：

```python
header = rows[0]
body = rows[1:]
```

header是一条字符串列表，body是多条记录组成的嵌套列表。header和body只是便于理解的变量名，不是关键字，也不是CSV工具自动创建的特殊对象。

本文件列顺序如下：

| 索引 | 列名 | 含义 |
| --- | --- | --- |
| 0 | id | 标识 |
| 1 | title | 游戏名称 |
| 2 | score | 评分 |
| 3 | game_type | 类型 |
| 4 | pulisher | 厂商信息，保留源拼写 |
| 5 | download | 下载量 |
| 6 | isAndroid | 平台标记，文本TRUE/FALSE |

先前简化三字段列表的score索引是1；完整文件的score索引是2。**列位置来自具体数据结构，不属于评分概念本身。**

### 代码与案例

In [16]:
header = rows[0]
body = rows[1:]
print("Column names:", header)
print("First data record:", body[0])
print("First two data records:", body[:2])
print("First score:", body[0][2])

Column names: ['id', 'title', 'score', 'game_type', 'pulisher', 'download', 'isAndroid']
First data record: ['206369', '人格解体', '8.8', '角色扮演', '厂商:\xa0喵法自然工作室', '71069', 'TRUE']
First two data records: [['206369', '人格解体', '8.8', '角色扮演', '厂商:\xa0喵法自然工作室', '71069', 'TRUE'], ['159465', '一念逍遥', '6.5', '角色扮演', '厂商:\xa0雷霆游戏', '1021860', 'TRUE']]
First score: 8.8


### 代码讲解、易错点与练习

body[0]是人格解体的记录，不再是列名；body[:2]得到两条数据。rows[:2]则包括表头和第一条记录。跳过表头只改变选取的记录范围，没有转换评分类型。

1. **K9.1** 已知明日方舟是第三条数据，用body取出它的title和score。解释为什么行索引与直接使用rows时不同。
2. **K9.2** 判断`len(header)`与`len(body)`分别表示什么；如果用len(rows)作为数据记录数，会多算什么？

**作答：** 行索引区别：____；三个长度的含义：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

## 10. for循环：语法、循环变量与执行顺序

### 理论知识

**循环让同一段代码重复执行。** for用于依次遍历可迭代对象中的元素。本章中列表和字符串都可以逐个提供元素，不需要先写出固定循环次数。

基本语法：

```python
for variable in iterable:
    statement_1
    statement_2
statement_after_loop
```

- for与in是语法组成部分；variable是当前元素使用的变量名。
- iterable是被遍历的对象；每轮提供一个元素。
- 冒号结束循环头，缩进语句组成循环体；本章统一四个空格。
- statement_after_loop没有缩进，等所有元素处理完成后才执行一次。

每轮的实际过程是：**取下一个元素 → 赋给循环变量 → 执行循环体**。下一轮把同一变量名绑定到另一个元素；它不会自动变成行号或序号。遍历空列表时，循环体一次也不执行。

### 代码：每轮处理一个名称

In [17]:
titles = ["人格解体", "一念逍遥", "明日方舟"]
for title in titles:
    print(title)
print("Finished titles")


人格解体
一念逍遥
明日方舟
Finished titles


### 代码讲解

循环变量title先后得到三个名称。缩进的print执行三次，最后一句执行一次。代码的书写顺序没有变，但循环会重复执行同一个代码块。

遍历body时，每个元素是一整行，row便是一条七字段列表。名称和评分分别使用row[1]、row[2]。只需要一个循环：外层取记录，列索引取固定字段，不必再循环所有列。

下面直接使用已读入的完整文件，只显示其前五条，避免大量输出掩盖结构。

In [18]:
for row in body[:5]:
    print(row[1], row[2])
print("Finished preview")

人格解体 8.8
一念逍遥 6.5
明日方舟 8.2
2048清 9.9
炉石传说 7
Finished preview


### 变量变化与易错点

| 轮次 | row代表 | row[1] | row[2] |
| --- | --- | --- | --- |
| 1 | 第一条七字段记录 | 人格解体 | '8.8' |
| 2 | 第二条七字段记录 | 一念逍遥 | '6.5' |
| 3 | 第三条七字段记录 | 明日方舟 | '8.2' |

row不是0、1、2。若把完成提示缩进，它会每轮重复；若把每行需要执行的print移到循环外，它只执行一次，通常只看到最后一次赋值的结果。

### 练习

1. **K10.1** 对body[:2]写一个循环，每行输出“游戏类型 名称”，循环后输出一次`Done`。先写出两轮row分别代表哪条记录，再运行核对。
2. **K10.2** 下方故障代码没有打印每个名称。预测输出，解释赋值与print的执行次数，然后修正。

```python
# Deliberate logic fault; keep it outside normal execution cells.
for row in body[:2]:
    name = row[1]
print(name)
```

**作答：** 每轮对象：____；故障输出预测/原因/修正：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

## 11. 类型转换与累加器：保存不断变化的总和

### 理论知识与语法

**累加器（accumulator）是用来保存逐步累计结果的变量。** 它是变量的一种使用方式，不是新的数据类型或Python关键字。

求和时需要两个步骤：

```python
total = 0.0             # 循环前建立起点
for value in values:
    total = total + value
```

右侧先读取旧total与当前value，计算结果，再重新赋给左侧total。也可以写`total += value`，但使用前total必须已初始化。本章先用完整写法看清过程。

如果value来自CSV，它通常是字符串，应先转换：`number = float(value)`。注意：

- 字符串加字符串是拼接，例如`"8.8" + "6.5"`得到文本`"8.86.5"`。
- 数值加字符串（如`0.0 + "8.8"`）会TypeError。
- float只能转换表示数值的值；对字段名`"score"`转换会ValueError，对一整行列表转换会TypeError。

### 代码：观察前三条评分的累计过程

In [19]:
score_texts = ["8.8", "6.5", "8.2"]
running_total = 0.0
for text in score_texts:
    number = float(text)
    running_total = running_total + number
    print("Current score:", number, "Running total:", running_total)

Current score: 8.8 Running total: 8.8
Current score: 6.5 Running total: 15.3
Current score: 8.2 Running total: 23.5


### 代码讲解

总和依次为8.8、15.3、23.5。这说明每轮不仅使用当前值，也保留之前各轮的结果；浮点输出偶尔有末位尾差，不改变这些数学关系。

把同一规则用于文件记录，只需将“取出评分文本”明确写为row[2]。下例仍只对原文件前三条做演示：

In [20]:
first_three_total = 0.0
for row in body[:3]:
    numeric_score = float(row[2])
    first_three_total = first_three_total + numeric_score
print("First-three total:", first_three_total)

First-three total: 23.5


### 易错点与练习

初始化只做一次，更新每轮做一次。如果把total=0.0放进循环体，之前的结果会不断被清空。重新运行整段统计代码时，也应重新执行初始化，避免延续上一次的总和。

1. **K11.1** 原文件第四、第五条评分为`"9.9"`、`"7"`。用一个小循环转换并累加，打印每轮总分，先手算再比较。
2. **K11.2** 下方代码每轮都重置total。预测最终值并解释为何不是两条评分之和，然后修正初始化位置。

```python
# Deliberate logic fault.
for text in ["9.9", "7"]:
    total = 0.0
    total = total + float(text)
print(total)
```

**作答：** 手算/程序对照：____；错误的状态变化：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

## 12. 平均值：总和与分母必须对应同一批数据

### 理论知识与公式

**算术平均值**等于参与计算的所有数值之和，除以这些数值的个数：

```text
平均值 = 数值总和 / 数值个数
```

这里每条记录贡献一次评分。分母应数参与求和的记录，而不是表头、字段数量或某个随手写入的常数。循环结束后才拥有完整总和，随后计算平均值。

```python
average = total / len(selected_rows)
```

前提是selected_rows非空；空列表的长度为0，直接相除会ZeroDivisionError。本章的原文件和所选样本均非空。处理任意空结果需要后续的条件判断。

### 代码：同一批数值的总和与数量

In [21]:
scores = [8.8, 6.5, 8.2]
three_score_total = 0.0
for score in scores:
    three_score_total = three_score_total + score
three_score_average = three_score_total / len(scores)
print("Mean of three scores:", three_score_average)

Mean of three scores: 7.833333333333333


### 代码讲解与原文件案例

total累计了scores中的三项，len(scores)也只数这三项，因此总和与分母对应。把同样结构迁移到CSV：从同一个原文件读取全部记录，抽取前五条计算样本平均分。

这五条原始评分为8.8、6.5、8.2、9.9、7，数学总和40.4，平均值8.08。下列完整例子展示读取、选择、转换、累计、除法各自的位置。它计算的是**前五条样本**，不是全表结果。

In [22]:
import csv

with open("mobile_game_info.csv", "r", encoding="utf-8-sig", newline="") as file:
    rows = list(csv.reader(file))
header = rows[0]
body = rows[1:]
selected_rows = body[:5]

sample_total = 0.0
for row in selected_rows:
    score = float(row[2])
    sample_total = sample_total + score
sample_average = sample_total / len(selected_rows)
print("Sample records:", len(selected_rows))
print("Sample total:", sample_total)
print("Sample mean:", sample_average)

Sample records: 5
Sample total: 40.4
Sample mean: 8.08


### 易错点与练习

不能用前五条的总和除以完整文件的记录数，也不能用全表总和除以七个字段的数量。结果在评分范围内只能说明“看起来可能合理”，还需要检查列、选取范围、总和和分母。

1. **K12.1** 已经只累计selected_rows的评分，却写`sample_total / len(rows)`。指出分母混入了哪些不应对应当前总和的行，并写出修正表达式。
2. **K12.2** 只计算原文件第四、第五条记录的平均分。先确定切片范围，再累计并用该子集的长度作分母；用手算验证。

**作答：** 分母问题：____；子集范围/手算/程序比较：____。

In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.

## 综合练习：独立计算完整文件的平均评分

使用同一份[mobile_game_info.csv](mobile_game_info.csv)。本题统计全部原始数据记录，名称重复也照常计入；不做去重、不筛平台、不按下载量加权。记录数不等于唯一游戏数量。原始数据的采集时间未随文件给出，不将结果描述为当前市场的最新平均评分。

### P1.1　读取与检查

独立写出读取代码，保留rows、header和body。打印表头、前五条记录和数据记录数；说明score列的位置、字段类型，以及为什么排除表头。

In [ ]:
# P1.1: Read and inspect the complete file here.

### P1.2　全表统计

用一个for循环处理所有body记录，转换score后累加。保留score_total和average_score；输出总评分、记录数和平均值。不把记录数或最终结果硬编码到代码中。

**作答：** 总分初始化位置与理由：____；实际循环范围：____；分母与统计含义：____。

In [ ]:
# P1.2: Calculate the full-file total and mean here.

### P1.3　验证与解释

从body中选取连续两条或三条记录，写明切片范围，先手算小计，再用代码计算同一子集的小计。不要把小样本手算值与完整文件总和直接比较。

重启内核，从头运行已完成的代码，确认读取与统计能复现。故障题的原错误片段保留在文字中，正常运行单元格只放修正代码。

**作答：** 所选记录/原始评分：____；手算小计：____；程序小计与比较：____；重启后的验证情况：____。

In [ ]:
# P1.3: Verify the selected subset and record your fresh-kernel observations.

## 本章总结

1. 列表是有序、可修改的序列；创建、取值、修改是不同操作。
2. 索引取一个元素，切片取一段列表；len返回数量，不是最后索引。
3. 嵌套列表分外层和内层，访问前先分清整表、记录和字段。
4. open返回文件对象，csv.reader解析记录，list收集成表；with管理文件关闭。
5. 表头描述字段，body保留数据；选择字段时根据当前表头确定索引。
6. for每轮重新给循环变量赋值，缩进控制重复执行的语句。
7. 评分先转成数值再累计；平均值的分母必须与求和范围一致。

保留完整文件的读取与统计代码。下次可以在同一批记录上增加条件判断，比较不同类型或平台的数据。

课后请打开 [chapter03_列表文件与循环_课后练习.ipynb](chapter03_列表文件与循环_课后练习.ipynb)，使用 [googleplaystore.csv](googleplaystore.csv) 独立完成 P1、P2（P3 选做）。两份数据的列名和列数不同，不能把课堂文件的下标直接套到课后文件上。
